packages/x-tree-view/src/internals/components/RichTreeViewItems.tsx

```diff
@@ -5,14 +5,148 @@ import { TreeItem, TreeItemProps } from '../../TreeItem';
 import { TreeViewItemId } from '../../models';
 import { TreeViewItemToRenderProps } from '../plugins/useTreeViewItems';
 
+const RichTreeViewItemsContext = React.createContext<
+  ((item: TreeViewItemToRenderProps) => React.ReactNode) | null
+>(null);
+
+if (process.env.NODE_ENV !== 'production') {
+  RichTreeViewItemsContext.displayName = 'RichTreeViewItemsProvider';
+}
+
+const areChildrenEqual = (childA: TreeViewItemToRenderProps, childB: TreeViewItemToRenderProps) => {
+  if (childA.itemId !== childB.itemId) {
+    return false;
+  }
+  if (childA.id !== childB.id) {
+    return false;
+  }
+  if (childA.label !== childB.label) {
+    return false;
+  }
+  if (childA.children.length !== childB.children.length) {
+    return false;
+  }
+  for (let i = 0; i < childA.children.length; i += 1) {
+    if (!areChildrenEqual(childA.children[i], childB.children[i])) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Logic copied from `fastObjectShallowCompare` but with a deep comparison for `props.children`
+const is = Object.is;
+const propsAreEqual = (a: WrappedTreeItemProps, b: WrappedTreeItemProps) => {
```

Comment:
I should be able to skip this custom function if `itemsToRender` was a string instead of an array.
I could just use `JSON.stringigy` but it's ugly as hell :laughing: 

I will make some additional work to see if I can store `itemsToRender` to render in the state (or in a cache that do not trigger a re-render, otherwise the virtualization scrolling would re-render the Tree View very often).